### DISCUSSION TITLE
Positional encoding for Audio Encoder and Text Decoder

### DISCUSSION THREAD

**Discussion by umbertocappellazzo** [openai/whisper#697] [Q&A] [answered]
Hi,
I was wondering why the positional encoding for the Audio Encoder is not learned during training phase (**register_buffer** used), whereas for the Text Encoder the pos encoding is defined as **nn.Parameter**, therefore it is learned. Is there any intuition for this? Other transformer implementations use **register_buffer** for both encoder and decoder, so I'm bit confused on the right choice to adhere to. Thank you!

**Comment by jongwook** [comment, marked as answer]
We used sinusoidal encoding for the encoder and learned positional encoding for the decoder. The sinusoidal encoding was proposed in the [transformer paper](https://arxiv.org/abs/1706.03762) because the authors "hypothesized it would allow the model to easily learn to attend by relative positions". In their experiments sinusoidal and learned embeddings didn't make much difference (Table 3, base vs. row E).

For language models (which is similar to Whisper's decoder), we empirically know that learned positional encoding works well, while we were more conservative in the audio encoder and used sinusoidal positional encoding. In any case, as shown in the transformer paper, the difference between sinusoidal vs learned positional encoding should be minor.

**Reply by raman-r-4978** [reply]
The way you have defined sinusoidal encodings [here](https://github.com/openai/whisper/blob/6dea21fd7f7253bfe450f1e2512a0fe47ee2d258/whisper/model.py#L53) is bit different than original transformer paper, where the positional embedding matrix consists of `sin` and `cos` in alternative positions whereas your implementation uses `torch.cat` to stack `sin vectors` and `cos vectors`. Any reason behind this?

**Reply by jongwook** [reply]
No particular reason. It won't have any different effects to the learning.

**Comment by umbertocappellazzo** [comment]
Thank you for your explanation. Now it's clear to me.
Sorry, I'd have another question, related to the decoder implementation. I see that the parameter **n_text_ctx** is fixed once and for all when the Whisper model is instantiated. This means that the text sequence length is the same for all samples, hence all text sequences are truncated or padded such that they match **n_text_ctx**, similar to what is done to the audio signal. Am I right?

**Reply by jongwook** [reply]
Yes. The model sees padded text sequences during training. A difference for the decoder is that it's used for autoregressive decoding, and it can stop when it sees the `<|endoftext|>` token.
